### PR TITLE
Start migrating API calls to Juju 2.0.

### PR DESCRIPTION
**Pull request by frankban**
Handle ModelManager calls, and also Model(Get|Info) Client calls.
Also improve createNev so that in theory it is able to create
new models in all provider types.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/4442/
Test PASSed.

File: jujugui/static/gui/src/app/app.js [hunk 1]
```diff
@@ -1719,6 +1720,7 @@ YUI.add('juju-gui', function(Y) {
       @param {String} password The password for the new environment.
     */
     switchEnv: function(socketUrl, username, password) {
+      console.log('switching to new socket URL:', socketUrl);
```
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: log left in.

File: jujugui/static/gui/src/app/components/env-switcher/env-switcher.js [hunk 2]
```diff
@@ -194,7 +194,7 @@ YUI.add('env-switcher', function() {
       // We need to coerce error types returned by JES vs JEM into one error.
       var err = data.err || error;
       if (err) {
-        console.log(err);
+        console.log('error creating new model:', err);
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/env-switcher/env-switcher.js]: this should really be surfaced to the user.

File: jujugui/static/gui/src/app/store/env/go.js [hunk 3]
```diff
@@ -2741,37 +2754,17 @@ YUI.add('juju-env-go', function(Y) {
...
-              config['maas-agent-name'] = data.config['maas-agent-name'];
-              break;
-            default:
-              // XXX frankban: add support for the remaining Juju providers.
-              intermediateCallback({
-                Error: ptype + ' provider is not supported yet'
-              });
-              return;
-          }
+          Object.keys(data.config).forEach((attr) => {
```
Comment (hatched) [inline jujugui/static/gui/src/app/store/env/go.js]: Nice cleanup

File: jujugui/static/gui/src/app/store/env/go.js [hunk 4]
```diff
@@ -2714,10 +2721,16 @@ YUI.add('juju-env-go', function(Y) {
           console.log('createEnv done: err:', data.Error);
         };
       }
+      var facade = 'ModelManager';
+      var request = 'CreateModel';
+      if (this.findFacadeVersion(facade) === null) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/store/env/go.js]: could you add a comment here as to why 'ModelManager' would not be found.

File: jujugui/static/gui/src/app/app.js (hunk 1 above)
Comment (frankban) [inline jujugui/static/gui/src/app/app.js]: Let's leave that for the time being. They will help debugging websocket stuff.

File: jujugui/static/gui/src/app/components/env-switcher/env-switcher.js (hunk 2 above)
Comment (frankban) [inline jujugui/static/gui/src/app/components/env-switcher/env-switcher.js]: Indeed, we should really return the error to the caller, I'll add an XXX.

File: jujugui/static/gui/src/app/store/env/go.js (hunk 3 above)
Comment (frankban) [inline jujugui/static/gui/src/app/store/env/go.js]: Thanks, hope it works.

File: jujugui/static/gui/src/app/store/env/go.js (hunk 4 above)
Comment (frankban) [inline jujugui/static/gui/src/app/store/env/go.js]: Done.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/4443/
Test PASSed.

**Comment by hatched** [comment]
:+1: 

QA :ok: ish :)

QA Notes:

- When I created a new model it didn't automatically switch to it for a very very long time. It should be showing the 'connecting' mask as soon as you create it as it switches. I believe you mentioned this on IRC already but the lack of 'connecting' mask is probably a bug as switching envs shows the mask (even if it takes a long long time to connect)
- Fonts are all bold, styling is off for some reason (we already knew about this)

**Comment by frankban** [comment]
Thank you for the review Jeff!
I also noticed the slow WebSocket conenction switching and I think it's related to closing the connection, and will investigate. Styling issues are not related to this branch.

**Comment by frankban** [comment]
Uros, I'll ship this to move forward, any comments from you will be tackled on next branch.
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**